src/mstore.erl

```diff
@@ -566,21 +602,28 @@ do_write(M=#mfile{offset=Offset, size=S, file=F, index=Idx},
          DataSize, Metric, Position, Value) when
       is_binary(Metric),
       is_binary(Value) ->
-    {M1, Base} =
-        case btrie:find(Metric, Idx) of
-            error ->
-                Pos = M#mfile.next,
-                Mx = M#mfile{next=Pos+1, index=btrie:store(Metric, Pos, Idx)},
-                file:write_file([M#mfile.name | ".idx"],
-                                <<(byte_size(Metric)):16/integer, Metric/binary>>,
-                                [read, append]),
-                {Mx, Pos*S*DataSize};
-            {ok, Pos} ->
-                {M,Pos*S*DataSize}
-        end,
-    P = Base+((Position - Offset)*DataSize),
-    R = file:pwrite(F, P, Value),
-    {R, M1}.
+    %% As part of a write operation, both the index and mstore files may be
+    %% updated. There is no atomicity gaurantees, and a fault may occur after
+    %% one of the writes have been processed. Writing data to the index before
+    %% writing to the mstore ensures that offsets are calculated so as not to
+    %% cause data to overlap. In other words, data loss is acceptable, but not
+    %% data corruption.
+    WritePoints = fun(Pos) ->
```

Comment:
I am not a big fan of the anonymous function here, it creates a closure over the variables and is really not needed it seems to makes the code harder to reason above since it no is a simple top to bottom logic.